Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java

```diff
@@ -197,6 +199,47 @@ public void saveTags(List<String> tagList) {
         mongoTemplate.save(new MongoProblemRepository.AllTags(tagList));
     }
 
+    @Override
+    public void updateProblemWithProvidedCodes(Problem problem, Language language, List<FileResource> providedCodes) {
+        String fileId = saveProvidedCodesAndGetFileId(problem.getId(), language, providedCodes);
+        updateProvidedCodesFileIdInProblem(problem, language, fileId);
+    }
+
+    private String saveProvidedCodesAndGetFileId(int problemId, Language language, List<FileResource> providedCodes) {
+        String fileName = format("%d-%s-provided.zip", problemId, language.toString());
+        ByteArrayInputStream zip = zipToStream(providedCodes);
+        return gridFsTemplate.store(zip, fileName).toString();
+    }
+
+    private void updateProvidedCodesFileIdInProblem(Problem problem, Language language, String fileId) {
+        problem.mayHaveLanguageEnv(language)
+                .ifPresentOrElse(langEnv -> {
+                    removeFileIdIfExist(langEnv.getProvidedCodesFileId());
+                    langEnv.setProvidedCodesFileId(fileId);
+                    updateLanguageEnv(problem.getId(), langEnv, language);
+                }, () -> {
+                    LanguageEnv langEnv = LanguageEnv.builder()
+                            .language(language)
+                            .providedCodesFileId(fileId)
+                            .build();
+                    problem.addLanguageEnv(langEnv);
+                    updateLanguageEnv(problem.getId(), langEnv, language);
+                });
+    }
```

Comment:
create an issue and add TODO comment for this:
Transaction: Operations in `updateProvidedCodesFileIdInProblem` should be atomic